qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,329 @@
+        has_basedir: bool = False
+    ) -> None:
+        self._has_basedir = has_basedir
+        self._in_progress = []  # type: typing.List[downloads.TempDownload]
+        self._done_count = 0
+        self._cache_path = str(data_dir / "adblock-cache.dat")
+        self._engine_factory = engine_factory
+        self._engine = engine_factory()
+
+    def _is_blocked(
+        self,
+        request_url: QUrl,
+        first_party_url: typing.Optional[QUrl] = None,
+        resource_type: typing.Optional[interceptor.ResourceType] = None,
+    ) -> bool:
+        """Check whether the given request is blocked."""
+        first_party_url = first_party_url
+        if first_party_url is not None and not first_party_url.isValid():
+            first_party_url = None
+
+        qtutils.ensure_valid(request_url)
+
+        if not config.get("content.blocking.adblock.enabled", url=first_party_url):
+            # Do nothing if adblocking is disabled.
+            return False
+
+        result = self._engine.check_network_urls(
+            request_url.toString(),
+            first_party_url.toString() if first_party_url else "",
+            resource_type_to_string(resource_type) if resource_type else "",
+        )
+
+        if not result.matched:
+            return False
+        if (result.exception is not None) and (result.important is None):
+            logger.debug(
+                "Excepting {} from being blocked by {} because of {}".format(
+                    request_url.toString(), result.filter, result.exception
+                )
+            )
+            return False
+        if _is_whitelisted_url(request_url):
+            logger.debug(
+                "Request to {} is whitelisted, thus not blocked".format(
+                    request_url.toString()
+                )
+            )
+            return False
+        return True
+
+    def filter_request(self, info: interceptor.Request) -> None:
+        """Block the given request if necessary."""
+        if self._is_blocked(info.request_url, info.first_party_url, info.resource_type):
+            logger.info(
+                "Request to {} blocked by ad blocker.".format(
+                    info.request_url.toString()
+                )
+            )
+            info.block()
+
+    def read_cache(self) -> None:
+        if os.path.isfile(self._cache_path):
+            logger.info("Loading cached adblock data: {}".format(self._cache_path))
+            self._engine.deserialize_from_file(self._cache_path)
+        else:
+            if (
+                config.val.content.blocking.adblock.lists
+                and not self._has_basedir
+                and config.val.content.blocking.adblock.enabled
+            ):
+                message.info("Run :brave-adblock-update to get adblock lists.")
+
+    def adblock_update(self) -> None:
+        """Update the adblock block lists."""
+        self._done_count = 0
+        self._engine = self._engine_factory()
```

Comment:
The `Engine` class doesn't have any method to "reset" its state, so I create a new engine instead. Such a method could be added either in my Python->Rust library, or upstream.

Reply:
Seems fine to have it here. I think it'd be nicer to create it directly rather than passing a factory method from the outside, but I'll comment on that separately.